src/python.rs

```diff
@@ -76,12 +110,12 @@ impl<'j> PythonParser<'j> {
                 Ok(StringCache::get(py, s.as_str()))
             }
             Peek::Array => {
-                let list = if let Some(peek_first) = self.parser.array_first()? {
+                let list = if let Some(peek_first) = tri!(self.parser.array_first(), PyList::empty(py)) {
                     let mut vec: SmallVec<[PyObject; 8]> = SmallVec::with_capacity(8);
-                    let v = self._check_take_value::<StringCache>(py, peek_first)?;
+                    let v = tri!(self._check_take_value::<StringCache>(py, peek_first), PyList::empty(py));
                     vec.push(v);
-                    while let Some(peek) = self.parser.array_step()? {
-                        let v = self._check_take_value::<StringCache>(py, peek)?;
+                    while let Some(peek) = tri!(self.parser.array_step(), PyList::new(py, vec)) {
+                        let v = tri!(self._check_take_value::<StringCache>(py, peek), PyList::new(py, vec));
                         vec.push(v);
                     }
                     PyList::new(py, vec)
```

Comment:
To avoid so many repeats to the `tri!` macro, what do you think of abstracting into a function? Could be something like:

```rust
                let mut vec: SmallVec<[PyObject; 8]> = SmallVec::with_capacity(8);
                if let Err(e) = self.parse_array(&mut list) {
                    if !self.allow_partial(&e) {
                        return Err(e);
                    }
                }
                Ok(PyList::new(py, vec).to_object(py))
```

... possibly without removing the special case to the empty list if it has a meaningful performance advantage.

Reply:
seems fair.

Reply:
done.